libyara/modules/macho.c

```diff
@@ -0,0 +1,1180 @@
+
+3. Neither the name of the copyright holder nor the names of its contributors
+may be used to endorse or promote products derived from this software without
+specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <yara/macho.h>
+#include <yara/endian.h>
+#include <yara/modules.h>
+#include <yara/mem.h>
+
+#define MODULE_NAME macho
+
+typedef struct _FAT_DATA {
+
+    // Single Mach-O file headers may be out of current block in fat Mach-O
+    // binaries. This structure stores auxiliary information for such cases.
+
+    size_t parsed_size; // Number of already parsed bytes
+    size_t file_count;  // Number of already parsed files
+
+    size_t count;    // Number of out of block offsets
+    size_t* offsets; // Mach-O files offsets
+
+} FAT_DATA;
+
+
+// Initialize FAT_MACHO_DATA structure.
+
+void macho_init_fat_data_storage(
+    FAT_DATA* storage)
+{
+  storage->parsed_size = 0;
+  storage->file_count = 0;
+  storage->count = 0;
+  storage->offsets = NULL;
+}
+
+
+// Check if file is for 32-bit architecture.
+
+int macho_is_32(uint8_t* magic)
+{
+  // Magic must be [CE]FAEDFE or FEEDFA[CE].
+  return magic[0] == 0xce || magic[3] == 0xce;
+}
+
+
+// Check if file is for big-endian architecture.
+
+int macho_is_big(uint8_t* magic)
+{
+  // Magic must be [FE]EDFACE or [FE]EDFACF.
+  return magic[0] == 0xfe;
+}
+
+
+// Check for Mach-O fat binary signature.
+
+int macho_is_fat_file_block(uint32_t* magic)
+{
+  return *magic == FAT_MAGIC || *magic == FAT_CIGAM;
+}
+
+
+// Check for Mach-O binary signatures.
+
+int macho_is_file_block(uint32_t* magic)
+{
+  return *magic == MH_MAGIC || *magic == MH_MAGIC_64 ||
+      *magic == MH_CIGAM || *magic == MH_CIGAM_64;
+}
+
+
+// Get length of string with limit (some strings are terminated by size limit).
+
+size_t macho_limited_strlen(char* string, size_t limit)
+{
+  size_t length = 0;
+  for(length = 0; string[length] && length < limit; length++);
+
+  return length;
+}
+
+
+// Convert virtual address to file offset. Segments have to be already loaded.
+
+int macho_rva_to_offset(
+    uint64_t address,
+    uint64_t* result,
+    YR_OBJECT* object
+    )
+{
+  uint64_t segment_count = get_integer(object, "number_of_segments");
```

Comment:
Is this problem even with exact-wide types (has to be exactly wide and in 2's complement)? If so, how to go around this? Via pointers - *(uint64_t*)&value? Because I need it to be unsigned (e.g. offset in file bigger that 2^32).